asset-transfer-basic/chaincode-javascript/lib/assetTransfer.js

```diff
@@ -0,0 +1,159 @@
+            },
+            {
+                ID: 'asset6',
+                Color: 'white',
+                Size: 15,
+                Owner: 'Michel',
+                AppraisedValue: 800,
+            },
+        ];
+
+        for (let i = 0; i < assets.length; i++) {
+            assets[i].docType = 'asset';
+            await ctx.stub.putState(assets[i].ID, Buffer.from(JSON.stringify(assets[i])));
+            console.info('Added <--> ', assets[i]);
+        }
+    }
+
+    // createAsset issues a new asset to the world state with given details.
+    async createAsset(ctx, id, color, size, owner, appraisedValue) {
+        const asset = {
+            ID: id,
+            Color: color,
+            Size: size,
+            Owner: owner,
+            AppraisedValue: appraisedValue,
+        };
+
+        await ctx.stub.putState(id, Buffer.from(JSON.stringify(asset)));
+    }
+
+    // readAsset returns the asset stored in the world state with given id.
+    async readAsset(ctx, id) {
+        const assetJSON = await ctx.stub.getState(id); // get the asset from chaincode state
+        if (!assetJSON || assetJSON.length === 0) {
+            throw new Error(`The asset ${id} does not exist`);
+        }
+
+        return assetJSON.toString();
+    }
+
+    // updateAsset updates an existing asset in the world state with provided parameters.
+    async updateAsset(ctx, id, color, size, owner, appraisedValue) {
+        const exists = await this.assetExists(ctx, id);
+        if (!exists) {
+            throw new Error(`The asset ${id} does not exist`);
+        }
+
+        // overwritting original asset with new asset
+        let updatedAsset = {
+            ID: id,
+            Color: color,
+            Size: size,
+            Owner: owner,
+            AppraisedValue: appraisedValue,
+        };
+
+        return ctx.stub.putState(id, Buffer.from(JSON.stringify(updatedAsset)));
+    }
+
+    // deleteAsset deletes an given asset from the world state.
+    async deleteAsset(ctx, id) {
+        const exists = await this.assetExists(ctx, id);
+        if (!exists) {
+            throw new Error(`The asset ${id} does not exist`);
+        }
+
+        return ctx.stub.deleteState(id);
+    }
+
+    // assetExists returns true when asset with given ID exists in world state.
+    async assetExists(ctx, id) {
+        const assetJSON = await ctx.stub.getState(id);
+        if (!assetJSON || assetJSON.length === 0) {
+            return false;
+        }
+        return true;
+    }
+
+    // transferAsset updates the owner field of asset with given id in the world state.
+    async transferAsset(ctx, id, newOwner) {
+        let assetString = await this.readAsset(ctx, id);
+
+        let asset = JSON.parse(assetString);
+        asset.Owner = newOwner;
+
+        await ctx.stub.putState(id, Buffer.from(JSON.stringify(asset)));
+    }
+
+    // getAllAssets returns all assets found in the world state.
+    async getAllAssets(ctx) {
```

Comment:
Suggestion:  Would suggest we don't include a 'getAll'  but rather always have it scoped down. 

getAll is fine when there are 10 records, but in reality it's going to be range query otherwise it'll take too long.

Reply:
This is a good point. 
Just currently `getAllAssets` are being used in multiple scripts. We can probably change this to get top 10 assets by assetID or something else.

Reply:
OK but let's not have a built-in range either. I think the chaincode needs to allow retrieving a range specified by the app and the app needs to be written to allow specifying that range via command line arguments. I dislike the fact that fabcar has this built-in and hope we can do better this time around!

Reply:
Yeah, Fabcar used to hardcode an arbitrary startKey and endKey, we just got that fixed to query the full open-ended range. I'd suggest since these samples are small to keep getAllAssets as full open-ended range query. And I think we can address Matthew's point by putting a comment in the code and tutorial that you can also choose to do scoped queries or paginated queries in production chaincodes (paginated queries are demonstrated in the more advanced ledger-queries variant of the sample chaincode).